File: software/examples/rust/example_threshold.rs
use std::{error::Error, io, thread};
use tinkerforge::{industrial_dual_0_20ma_bricklet::*, ip_connection::IpConnection};

const HOST: &str = "localhost";
const PORT: u16 = 4223;
const UID: &str = "XYZ"; // Change XYZ to the UID of your Industrial Dual 0-20mA Bricklet.

fn main() -> Result<(), Box<dyn Error>> {
    let ipcon = IpConnection::new(); // Create IP connection.
    let id020 = IndustrialDual020maBricklet::new(UID, &ipcon); // Create device object.

    ipcon.connect((HOST, PORT)).recv()??; // Connect to brickd.
                                          // Don't use device before ipcon is connected.

    // Get threshold receivers with a debounce time of 10 seconds (10000ms).
    id020.set_debounce_period(10000);

    // Create receiver for current reached events.
    let current_reached_receiver = id020.get_current_reached_receiver();

    // Spawn thread to handle received events. This thread ends when the `id020` object
    // is dropped, so there is no need for manual cleanup.
    thread::spawn(move || {
        for current_reached in current_reached_receiver {
            println!("Sensor: {}", current_reached.sensor);
            println!("Current: {} mA", current_reached.current as f32 / 1000000.0);
            println!();
        }
    });

    // Configure threshold for current (sensor 1) "greater than 10 mA".
    id020.set_current_callback_threshold(1, '>', 10 * 1000000, 0);

    println!("Press enter to exit.");
    let mut _input = String::new();
    io::stdin().read_line(&mut _input)?;
    ipcon.disconnect();
    Ok(())
}

File: software/examples/go/example_threshold.go
package main

import (
	"fmt"
	"github.com/tinkerforge/go-api-bindings/industrial_dual_0_20ma_bricklet"
	"github.com/tinkerforge/go-api-bindings/ipconnection"
)

const ADDR string = "localhost:4223"
const UID string = "XYZ" // Change XYZ to the UID of your Industrial Dual 0-20mA Bricklet.

func main() {
	ipcon := ipconnection.New()
	defer ipcon.Close()
	id020, _ := industrial_dual_0_20ma_bricklet.New(UID, &ipcon) // Create device object.

	ipcon.Connect(ADDR) // Connect to brickd.
	defer ipcon.Disconnect()
	// Don't use device before ipcon is connected.

	// Get threshold receivers with a debounce time of 10 seconds (10000ms).
	id020.SetDebouncePeriod(10000)

	id020.RegisterCurrentReachedCallback(func(sensor uint8, current int32) {
		fmt.Printf("Sensor: %d\n", sensor)
		fmt.Printf("Current: %f mA\n", float64(current)/1000000.0)
		fmt.Println()
	})

	// Configure threshold for current (sensor 1) "greater than 10 mA".
	id020.SetCurrentCallbackThreshold(1, '>', 10*1000000, 0)

	fmt.Print("Press enter to exit.")
	fmt.Scanln()

}
